src/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -146,28 +140,18 @@ public void NextProperty()
         {
             EndProperty();
 
-            if (PropertyEnumeratorActive)
+            int len = JsonClassInfo.PropertyCacheArray.Length;
+            if (++PropertyEnumeratorIndex >= len)
             {
-                int len = JsonClassInfo.PropertyCacheArray.Length;
-                if (PropertyEnumeratorIndex < len)
+                if (PropertyEnumeratorIndex > len)
```

Comment:
I am not fully understanding this logic. Can you explain what we are trying to do here?

Reply:
The extension property is always the last property in `PropertyCacheArray` so this is checking to see if the extension property is writing or whether it is finished.

The extension property which is different than other properties because although it is a dictionary property we want to serialize each element in the dictionary as a property of the POCO instead of a dictionary element thus we need some extra state to track whether it is being serialized.